hutch_python/epics_arch.py

```diff
@@ -93,13 +116,138 @@ def create_arch_file(experiment, hutch=None, path=None, dry_run=False):
             file_path = path
         elif hutch:
             file_path = EPICS_ARCH_FILE_PATH.format(hutch.lower())
+        elif cds_items:
+            pull_cds_data(experiment, cds_items)
+            return
+        elif link:
+            update_file(exp_name=experiment, path=EPICS_ARCH_FILE_PATH.format(experiment[0:3]))
+            create_softlink(experiment)
+            return
         else:
             file_path = EPICS_ARCH_FILE_PATH.format(experiment[0:3])
-        create_file(exp_name=experiment, path=file_path)
+        update_file(exp_name=experiment, path=file_path)
     elif dry_run:
         print_dry_run(experiment)
 
 
+def pull_cds_data(exp, run):
+    logger.debug("in client")
+    pull_cds_items(exp, run)
+
+
+def create_softlink(experiment):
+    logger.debug("in softlink")
+    # remove the old soft link and add a new one (update), *THIS HAS NOT BEEN TESTED YET*
+    # this removes the softlink in the /cds/group/pcds/dist/pds/{}/misc/
+    # rm_result = subprocess.run(['unlink', EPICS_ARCH_FILE_PATH.format(experiment[0:3]) + 'epicsArch_' + experiment[0:3].upper() + '_exp_specific.txt'])
+
+    # This adds a new softlink in /cds/group/pcds/dist/pds/{}/misc/
+    subprocess.run(['ln', '-sf', EPICS_ARCH_FILE_PATH.format(experiment[0:3]) + 'epicsArch_' + experiment + '.txt', EPICS_ARCH_FILE_PATH.format(experiment[0:3]) + 'epicsArch_' + experiment[0:3].upper() + '_exp_specific.txt'])
+
+
+def check_for_duplicates(qs_data, af_data):
+
+    """
+    Check for duplicate PVs in the questionnaire, the code already throws a warning for duplicate aliases.
+    If duplicates (PV or Alias) are found in the questionnaire throw error and prompt user to fix and re-run. If they are found in the epicsArch file then step through each match and update accordingly.
+
+    Parameters
+    ----------
+    qs_data : list
+    af_data : list
+
+    Examples
+    --------
+    >>> epicsarch-qs xpplv6818 --dry-run
+
+    Returns
+    -------
+    updatea_arch_list : list
+        Updated list containing sorted alias, PVs.
+
+    """
+
+    """
+    Part 1: Parse Data from the questionnaire and the archfile
+    Part 2: Check the questionnaire for pv duplicates
+    """
+
+    # PART 1
+
+    # Convert lists to dictionaries to sort as a key - value pair while also removing any whitespice in the aliases.
+
+    # Questionnaire Data, removing whitespaces and newline chars
+    qsDict = dict(zip(qs_data[::2], qs_data[1::2]))
+    qsDict = {k.replace(" ", ""): v for k, v in qsDict.items()}
+    qsDict = {k.replace("\n", ""): v for k, v in qsDict.items()}
+    qsDict = {k: v.replace(" ", "") for k, v in qsDict.items()}
+    qsDict = {k: v.replace("\n", "") for k, v in qsDict.items()}
+    sorted_qsDict = dict(sorted(qsDict.items()))
+
+    # If the archfile is not empty then clean it if not ,skip
+    if len(af_data) > 0:
```

Comment:
Nitpick: A consequence here is that `afDict` and `sorted_afDict` will not be defined if there is no archfile data. This could cause issues later in the function. I suggesting adding an `else` clause where you set these to empty dictionaries.

edit: I see later that you return early in these cases, so I'm downgrading this to nitpick